### PR TITLE
homed: add concept for "maximizing" home dirs

### PR DESCRIPTION
**Pull request by poettering**
A previous PR added the ability to "minimize" home dirs, i.e. shrink them to their minimal size taking disk usage/file system constraints into account. This one adds the opposite, it allows growing the disk size to the max, and handling growing gracefully: i.e. we'll grow to the maximum fallocate() allows us to even if that's smaller than what we got asked for.

### CODE REVIEW

**Comment by poettering** [comment]
Split out of #21253

File: src/home/homework.h [hunk 1]
```diff
@@ -63,6 +63,7 @@ typedef enum HomeSetupFlags {
         HOME_SETUP_RESIZE_MINIMIZE             = 1 << 3, /* Shrink to minimal size */
         HOME_SETUP_RESIZE_DONT_GROW            = 1 << 4, /* If the resize would grow, gracefully terminate operation */
         HOME_SETUP_RESIZE_DONT_SHRINK          = 1 << 5, /* If the resize would shrink, gracefully terminate operation */
+        HOME_SETUP_RESIZE_DONT_UNDO            = 1 << 6, /* Leave loopback/DM device context open after succesful operation */
```
Comment (bluca) [inline src/home/homework.h]: /succesful/successful/

File: src/home/homework-luks.c [hunk 2]
```diff
@@ -2851,6 +2882,77 @@ static int resize_fs_loop(
...
+        uint64_t current_image_size;
+        unsigned n_iterations = 0;
+        int r;
+
+        assert(h);
+        assert(setup);
+        assert(setup->image_fd >= 0);
+
+        /* A bisection loop trying to find the closest size to what the user asked for. (Well, we bisect like
+         * this only when we *grow* the image — if we shrin the image then there's no need to bisect.) */
```
Comment (bluca) [inline src/home/homework-luks.c]: /shrin/shrink/

File: src/home/homework-luks.c [hunk 3]
```diff
@@ -2851,6 +2882,77 @@ static int resize_fs_loop(
...
+                        log_debug("Resizing from %s to %s via truncation worked successfully.", FORMAT_BYTES(old_image_size), FORMAT_BYTES(try_image_size));
+                        current_image_size = try_image_size;
+                        break; /* there's no point in the bisection logic if this was plain truncation and
+                                * not allocation, let's exit immediately. */
+                }
+
+                if (new_image_size < old_image_size) /* If we are shrinking we are done after one iteration */
+                        break;
+
+                /* If we are growing then let's adjust our bisection boundries and try again */
```
Comment (bluca) [inline src/home/homework-luks.c]: /boundries/boundaries/

File: src/home/homectl.c [hunk 4]
```diff
@@ -1763,6 +1763,32 @@ static int passwd_home(int argc, char *argv[], void *userdata) {
...
+        else if (streq(t, "max"))
+                *ret = UINT64_MAX-1;  /* Largest size that isn't UINT64_MAX special marker */
+        else {
+                uint64_t ds;
+
+                r = parse_size(t, 1024, &ds);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to parse disk size parameter: %s", t);
+
+                if (ds >= UINT64_MAX) /* UINT64_MAX has special meaning for us ("dont change"), refuse */
```
Comment (bluca) [inline src/home/homectl.c:1783]: UINT64_MAX-1 is special too, no? shouldn't that be refused too?

**Review by bluca — APPROVED** [review]
lgtm, just some nits

File: src/home/homectl.c (hunk 4 above)
Comment (poettering) [reply src/home/homectl.c:1783]: No it's not special really. It's the largest of the non-special values, and that makes it a bit special if you so will ;-)

So if you grow to the home dir to the max you can just specify the largest valid size and we'll try our best to get there.

**Comment by poettering** [comment]
fixed the typos, upgraded label.

**Pull request merged**